File: src/strata/memory/_helpers/bootstrap.py
"""Safely create canonical source directories for enabled repository memory."""

from __future__ import annotations

import os
import stat
from pathlib import Path

from strata.memory._helpers.native_operations import inspect_sources
from strata.memory.constants import (
    MEMORY_BOOTSTRAP_FILENAME,
    MEMORY_DATABASE_DIRECTORY,
    MEMORY_DIRECTORIES,
    MEMORY_GITIGNORE_ENTRY,
)
from strata.memory.exceptions import MemoryOperationError
from strata.memory.models import MemoryIndexSummary

_GITIGNORE_NAME: str = ".gitignore"
_GITIGNORE_BLOCK: bytes = f"# Strata\n{MEMORY_GITIGNORE_ENTRY}\n".encode()
_FILE_MODE: int = 0o644


def bootstrap_memory(repository_root: Path) -> None:
    """Validate existing sources and create missing canonical Memory state once."""

    marker_path: Path = repository_root / MEMORY_DATABASE_DIRECTORY / MEMORY_BOOTSTRAP_FILENAME
    if marker_path.is_file() and _has_complete_structure(repository_root=repository_root):
        _ensure_gitignore(repository_root=repository_root)
        return
    summary: MemoryIndexSummary = inspect_sources(repository_root)
    diagnostics: int = (
        summary.source_diagnostic_count
        + summary.corpus_diagnostic_count
        + summary.graph_diagnostic_count
    )
    if diagnostics:
        raise MemoryOperationError(
            "Existing .ai content is not canonical and will not be migrated automatically; "
            "migrate it manually before using memory."
        )
    _ensure_gitignore(repository_root=repository_root)
    for relative_path in MEMORY_DIRECTORIES:
        (repository_root / relative_path).mkdir(parents=True, exist_ok=True)
    marker_path.parent.mkdir(parents=True, exist_ok=True)
    marker_path.touch(exist_ok=True)


def _has_complete_structure(*, repository_root: Path) -> bool:
    return all((repository_root / relative_path).is_dir() for relative_path in MEMORY_DIRECTORIES)


def _ensure_gitignore(*, repository_root: Path) -> None:
    path: Path = repository_root / _GITIGNORE_NAME
    flags: int = os.O_RDWR | os.O_APPEND | getattr(os, "O_NOFOLLOW", 0)
    try:
        descriptor: int = os.open(path, flags)
    except FileNotFoundError:
        _create_gitignore(path=path)
        return
    except OSError as error:
        raise MemoryOperationError(f"Memory bootstrap could not open {path}: {error}") from error
    try:
        metadata: os.stat_result = os.fstat(descriptor)
        if not stat.S_ISREG(metadata.st_mode):
            raise MemoryOperationError(
                f"Memory bootstrap requires a regular root gitignore: {path}"
            )
        content: bytes = os.read(descriptor, metadata.st_size)
        if MEMORY_GITIGNORE_ENTRY.encode() in content.splitlines():
            return
        separator: bytes = b"" if not content or content.endswith(b"\n") else b"\n"
        _write_all(descriptor=descriptor, content=separator + _GITIGNORE_BLOCK)
        os.fsync(descriptor)
    except OSError as error:
        raise MemoryOperationError(f"Memory bootstrap could not update {path}: {error}") from error
    finally:
        os.close(descriptor)


def _create_gitignore(*, path: Path) -> None:
    flags: int = os.O_WRONLY | os.O_CREAT | os.O_EXCL | getattr(os, "O_NOFOLLOW", 0)
    try:
        descriptor: int = os.open(path, flags, _FILE_MODE)
    except OSError as error:
        raise MemoryOperationError(f"Memory bootstrap could not create {path}: {error}") from error
    try:
        _write_all(descriptor=descriptor, content=_GITIGNORE_BLOCK)
        os.fsync(descriptor)
    except OSError as error:
        raise MemoryOperationError(f"Memory bootstrap could not write {path}: {error}") from error
    finally:
        os.close(descriptor)


def _write_all(*, descriptor: int, content: bytes) -> None:
    remaining: memoryview = memoryview(content)
    while remaining:
        written: int = os.write(descriptor, remaining)
        remaining = remaining[written:]

File: crates/fensu-memory/src/source/helpers/bootstrap.rs
use std::fs::{self, File, OpenOptions};
use std::io::{Read, Seek, SeekFrom, Write};
use std::path::{Path, PathBuf};

#[cfg(unix)]
use std::os::unix::fs::OpenOptionsExt;

use crate::engine::main::summarize_memory::summarize_memory;

const DATABASE_DIRECTORY: &str = ".fensu/memory";
const BOOTSTRAP_FILE: &str = ".bootstrapped";
const GITIGNORE_FILE: &str = ".gitignore";
const GITIGNORE_ENTRY: &str = ".fensu/memory/";
const GITIGNORE_BLOCK: &[u8] = b"# Fensu\n.fensu/memory/\n";
const MEMORY_DIRECTORIES: &[&str] = &[
    ".ai/tasks/not-started",
    ".ai/tasks/in-progress",
    ".ai/tasks/completed",
    ".ai/tasks/cancelled",
    ".ai/tasks/superseded",
    ".ai/knowledge/repo/notes",
    ".ai/knowledge/repo/decisions",
    ".ai/knowledge/repo/skills",
    ".ai/_archive/tasks/completed",
    ".ai/_archive/tasks/cancelled",
    ".ai/_archive/tasks/superseded",
    ".ai/_archive/knowledge/repo/notes",
    ".ai/_archive/knowledge/repo/decisions",
    ".ai/_archive/knowledge/repo/skills",
];

pub(crate) fn bootstrap(repository_root: &Path) -> Result<(), String> {
    let marker_path = repository_root
        .join(DATABASE_DIRECTORY)
        .join(BOOTSTRAP_FILE);
    if marker_path.is_file() && has_complete_structure(repository_root) {
        ensure_gitignore(repository_root)?;
        return Ok(());
    }
    let summary = summarize_memory(repository_root);
    let diagnostic_count = summary.source_diagnostic_count
        + summary.corpus_diagnostic_count
        + summary.graph_diagnostic_count;
    if diagnostic_count > 0 {
        return Err(
            "Existing .ai content is not canonical and will not be migrated automatically; migrate it manually before using memory."
                .to_owned(),
        );
    }
    ensure_gitignore(repository_root)?;
    for relative_path in MEMORY_DIRECTORIES {
        let path = repository_root.join(relative_path);
        fs::create_dir_all(&path).map_err(|error| {
            format!(
                "Memory bootstrap could not create directory {}: {error}",
                path.display()
            )
        })?;
    }
    let parent = marker_path
        .parent()
        .ok_or_else(|| "Memory bootstrap marker has no parent directory.".to_owned())?;
    fs::create_dir_all(parent).map_err(|error| {
        format!(
            "Memory bootstrap could not create directory {}: {error}",
            parent.display()
        )
    })?;
    OpenOptions::new()
        .write(true)
        .create(true)
        .truncate(false)
        .open(&marker_path)
        .map_err(|error| {
            format!(
                "Memory bootstrap could not create {}: {error}",
                marker_path.display()
            )
        })?;
    Ok(())
}

fn has_complete_structure(repository_root: &Path) -> bool {
    MEMORY_DIRECTORIES
        .iter()
        .all(|relative_path| repository_root.join(relative_path).is_dir())
}

fn ensure_gitignore(repository_root: &Path) -> Result<(), String> {
    let path = repository_root.join(GITIGNORE_FILE);
    match open_existing_gitignore(&path) {
        Ok(mut file) => update_gitignore(&mut file, &path),
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => create_gitignore(&path),
        Err(error) => Err(format!(
            "Memory bootstrap could not open {}: {error}",
            path.display()
        )),
    }
}

fn open_existing_gitignore(path: &Path) -> std::io::Result<File> {
    let mut options = OpenOptions::new();
    options.read(true).write(true).append(true);
    #[cfg(unix)]
    options.custom_flags(libc::O_NOFOLLOW);
    options.open(path)
}

fn update_gitignore(file: &mut File, path: &Path) -> Result<(), String> {
    let metadata = file.metadata().map_err(|error| {
        format!(
            "Memory bootstrap could not inspect {}: {error}",
            path.display()
        )
    })?;
    if !metadata.is_file() {
        return Err(format!(
            "Memory bootstrap requires a regular root gitignore: {}",
            path.display()
        ));
    }
    file.seek(SeekFrom::Start(0)).map_err(|error| {
        format!(
            "Memory bootstrap could not update {}: {error}",
            path.display()
        )
    })?;
    let mut content = Vec::new();
    file.read_to_end(&mut content).map_err(|error| {
        format!(
            "Memory bootstrap could not update {}: {error}",
            path.display()
        )
    })?;
    if content.split(|byte| *byte == b'\n').any(|line| {
        line.strip_suffix(b"\r") == Some(GITIGNORE_ENTRY.as_bytes())
            || line == GITIGNORE_ENTRY.as_bytes()
    }) {
        return Ok(());
    }
    if !content.is_empty() && !content.ends_with(b"\n") {
        file.write_all(b"\n")
            .map_err(|error| gitignore_write_error(path, error))?;
    }
    file.write_all(GITIGNORE_BLOCK)
        .map_err(|error| gitignore_write_error(path, error))?;
    file.sync_all()
        .map_err(|error| gitignore_write_error(path, error))?;
    Ok(())
}

fn create_gitignore(path: &Path) -> Result<(), String> {
    let mut options = OpenOptions::new();
    options.write(true).create_new(true);
    #[cfg(unix)]
    {
        options.mode(0o644).custom_flags(libc::O_NOFOLLOW);
    }
    let mut file = options.open(path).map_err(|error| {
        format!(
            "Memory bootstrap could not create {}: {error}",
            path.display()
        )
    })?;
    file.write_all(GITIGNORE_BLOCK)
        .map_err(|error| gitignore_write_error(path, error))?;
    file.sync_all()
        .map_err(|error| gitignore_write_error(path, error))?;
    Ok(())
}

fn gitignore_write_error(path: &Path, error: std::io::Error) -> String {
    format!(
        "Memory bootstrap could not write {}: {error}",
        PathBuf::from(path).display()
    )
}
